celeritas-project/celeritas
# Add GitHub actions to CI

Since our CI is down after almost a week, we've got to get some testing capability back online. This restores all the original build types and builds them through GitHub's runners using the original CI images.

Because the runners don't have GPUs, this builds the code and runs tests with `CELER_DISABLE_DEVICE`.
-  I added support for *runtime* skipping of device tests in googletest: if the device is unavailable, any test with `[Dd]evice` in the name gets skipped.
- The `app` testing checks for the environment variable at configure time, and it disables the GPU tests by default if found.
- To support runtime device-disabling for the hit manager (which uses pinned memory for the hits even in host-only mode) I have the `PinnedAllocator` call `Device::num_devices` to select between native (non-pinned) and device API mallocs.

The build names have been changed:
| Image | Old | New |
| ----- | --- | --- |
| ci-centos-rocm | asan      | asan-orange |
| ci-centos-rocm | minitest  | debug-orange |
| ci-centos-rocm | float     | debug-orange-float |
| ci-centos-rocm | geant4    | reldeb-geant4 |
| ci-centos-rocm | ndebug    | ndebug-orange |
| ci-ubuntu-cuda | full-novg        | debug-orange |
| ci-ubuntu-cuda | full-novg-ndebug | ndebug-orange |
| ci-ubuntu-cuda | vecgeom-reldeb   | reldeb-vecgeom |
| ci-ubuntu-cuda | vecgeom-tests    | debug-vecgeom |
| ci-ubuntu-cuda | vecgeom-demos    | ndebug-vecgeom |

I've also adjusted the cmake presets so that "ndebug" prefers static libraries. This doesn't work with ROOT (PCM loading looks only in bin, not lib) so I added cmake code to check for that. Thinking it might've been an omission in the scoped error handler, I updated the exception initalizer to try to load the PCM file for static builds.

TO DO in subsequent PRs:
- Fix the issues with examples
  * asan build needs `-fsanitize=address` in downstream example
  * rocm+ndebug (static libs) fails to propagate the HIP library linkage
  * SimpleOffload::Build crashes because MTLogger is called before run manager completes initialization
  * VecGeom build on Ubuntu requires `-Wl,--no-as-needed` which we provide to Celeritas via `CMAKE_{SHARED|EXE}_LINKER_FLAGS`: maybe @pcanal 's RDC code needs to propagate these? Should this be vecgeom's responsibility?
- Use a shallow clone and `git describe` or some better way of fetching tags
- Improve docker images
  * Add ccache to the build image and use it
  * Avoid custom user ID; GitHub requires root
  * Install all supported geant4 versions for a small build matrix
  * Fix entry point so that we don't have to manually source `/etc/profile` and re-export variables
- Add separate build matrix for geant4 versions (to test minimum and maximum supported)
- Add nightly minimal build matrix for various compilers on raw hardware (earliest GCC through latest, multiple clang versions)
- Add automatic clang formatting and/or fix?
- Add documentation
